src/dividat-driver/firmware/cli.go

```diff
@@ -0,0 +1,76 @@
+package firmware
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	"github.com/dividat/driver/src/dividat-driver/service"
+)
+
+// Command-line interface to running a firmware update
+func Command(flags []string) {
+	updateFlags := flag.NewFlagSet("update", flag.ExitOnError)
+	imagePath := updateFlags.String("i", "", "Firmware image path")
+	configuredAddr := updateFlags.String("a", "", "Senso address (optional)")
+	sensoSerial := updateFlags.String("s", "", "Senso serial (optional)")
+	updateFlags.Parse(flags)
+
+	if *imagePath == "" {
+		flag.PrintDefaults()
+		return
+	}
+	file, err := os.Open(*imagePath)
+	if err != nil {
+		fmt.Printf("Could not open image file: %v\n", err)
+		os.Exit(1)
+	}
+
+	onProgress := func(progressMsg string) {
+		fmt.Println(progressMsg)
+	}
+
+	if *sensoSerial != "" {
+		err = UpdateBySerial(context.Background(), *sensoSerial, file, onProgress)
+	} else if *configuredAddr != "" {
+		err = updateByAddress(context.Background(), *configuredAddr, file, onProgress)
+	} else {
+		err = updateByDiscovery(context.Background(), file, onProgress)
+	}
+
+	if err != nil {
+		fmt.Println()
+		fmt.Printf("Update failed: %v \n", err)
+		os.Exit(1)
+	}
+}
+
+// The following functions are only used when updating the firmware via the command line.
+// This is why they are private, and not part of the main module.
+
+func updateByAddress(ctx context.Context, address string, image io.Reader, onProgress OnProgress) error {
+	onProgress(fmt.Sprintf("Using specified address %s", address))
+	match := service.Find(ctx, 15*time.Second, service.AddressFilter(address))
```

Comment:
You're right, this slipped my attention. After browsing some sysadmin forums I get the impression it's not uncommon to block mDNS traffic. Since we allow people to configure Play to use a fixed address, I think we should extend this to the firmware upgrade as well. We could make `updateByAddress` assume that it's only used where mDNS is not available (ie. when Play is configured with a static address):

```go
func updateByAddress(ctx context.Context, address string, image io.Reader, onProgress OnProgress) error {
	onProgress(fmt.Sprintf("Using specified address %s", address))
	err := sendDfuCommand(address, controllerPort, onProgress)
	if err == nil {
		waitForTFTP(onProgress)
	} else {
		onProgress("Could not send DFU command. Senso may already be in bootloader mode.")
	}
	return putTFTP(address, tftpPort, image, onProgress)
}
```

This requires that a static DHCP lease is configured for the Senso, otherwise the Senso could potentially get stuck in bootloader mode if its address changes when it reboots. Hopefully clients with a static address configuration would set this up as well.

Reply:
> This requires that a static DHCP lease is configured for the Senso, otherwise the Senso could potentially get stuck in bootloader mode if its address changes when it reboots. Hopefully clients with a static address configuration would set this up as well.

Yes, we need to assume a statically configured address in this case. A Senso in normal mode could self-assign a new address each time, if no DHCP lease is given. So there should be a stable assignment mechanism, usually based on MAC.

> Since we allow people to configure Play to use a fixed address, I think we should extend this to the firmware upgrade as well.

This reasoning makes sense, but I think it equally makes sense to only support the mDNS case for firmware updates, because

- it is used by 99 % of installations,
- it helps us provide a robust update mechanism through better insight into peer state,
- we can focus our resources on making this one mechanism work well.

It would still be reasonable to say "normal operation is supported without mDNS, for updating attach Senso directly to Play Computer".

Reply:
The implementation of updating by address seems quite straightforward to me so I was tempted to keep it, but seeing that the overwhelming majority of people uses mDNS, I'm convinced to only offer updating via serial for now. Let's see how well this one mechanism will work under real-world conditions. We can always re-add updating by address if it proves to be relevant, but it would be harder to remove later if some people come to rely on it.

[Remove firmware update by fixed address](https://github.com/dividat/driver/pull/127/commits/8743fda024225174829514e81639e00105dced06)